### PR TITLE
Add ADMM subproblem solvers

### PR DESCRIPTION
**Pull request by bwohlberg**
Add ADMM x-step solvers for problems with a sum-of-convolutions type forward operator and a squared ℓ2 data fidelity term, partially addressing #232. Numerous other miscellaneous changes.

### CODE REVIEW

**Review by Michael-T-McCann — APPROVED** [review]
Cool, especially the diffusercam example. See a couple minor comments. I committed two typo fixes.

File: examples/scripts/diffusercam_tv_admm.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+plot.imview(xms, cmap=cmap, cbar=True, title="Recovered Volume")
+
+
+"""
+Plot convergence statistics.
+"""
+fig, ax = plot.subplots(nrows=1, ncols=2, figsize=(12, 5))
+plot.plot(
+    hist.Objective,
+    title="Objective function",
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/diffusercam_tv_admm.py:175]: I tried that. It doesn't help. One would have to plot the objective minus the minimum value to really see a benefit.

File: examples/scripts/sparsecode_conv_admm.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+
+"""
+Define loss function and regularization. Note the use of the
+$\ell_1 - \ell_2$ norm, which has been found to provide slightly better
+performance than the $\ell_1$ norm in this type of problem
+:cite:`wohlberg-2021-psf`.
+"""
+f = SquaredL2Loss(y=y, A=A)
+g0 = λ * DiffL1L2Norms()
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/sparsecode_conv_admm.py:92]: Yes, good point. Will fix shortly. On a related note, any thoughts on the name `DiffL1L2Norm`? It's not great, but I couldn't think of anything better.

File: examples/scripts/sparsecode_conv_admm.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+fig.show()
+
+
+"""
+Show test image and reconstruction from recovered coefficient maps.
+"""
+fig, ax = plot.subplots(nrows=1, ncols=2, figsize=(12, 6))
+plot.imview(y, title="Test image", cmap=plot.cm.gist_heat_r, fig=fig, ax=ax[0])
+plot.imview(A(x1), title="Reconstructed image", cmap=plot.cm.gist_heat_r, fig=fig, ax=ax[1])
+fig.show()
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/sparsecode_conv_admm.py:141]: The notebooks had not been run in the corresponding `scico-data` branch. Fixed now.

File: examples/scripts/sparsecode_conv_md_admm.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+# with the package.
+
+r"""
+Convolutional Sparse Coding with Mask Decoupling (ADMM)
+=======================================================
+
+This example demonstrates the solution of a convolutional sparse coding
+problem
+
+  $$\mathrm{argmin}_{\mathbf{x}} \; \frac{1}{2} \Big\| \mathbf{y} -
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/sparsecode_conv_md_admm.py:14]: I've been using (1/2) when the following element (e.g. open norm symbol) is small, and \frac{1}{2} when it's large enough that the frac is not of greater height. Your proposed policy is simpler, and not unreasonable, but the the one I've been using looks better, IMO.

File: examples/scripts/sparsecode_conv_md_admm.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+plot.imview(x1[2], cmap=plot.cm.Blues, fig=fig, ax=ax[1, 2])
+fig.tight_layout()
+fig.show()
+
+
+"""
+Show test image and reconstruction from recovered coefficient maps. Note
+the absence of the wrap-around effects at the boundary that can be seen
+in the corresponding images in the [related example](sparsecode_conv_admm.rst).
+"""
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/sparsecode_conv_md_admm.py:162]: These two examples have very similar run times. For bigger problems the MD form may be a bit more expensive, and it's often a bit trickier to tune, but I can't think of any succinct description that would be appropriate here.

File: examples/scripts/sparsecode_conv_admm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Michael-T-McCann) [reply examples/scripts/sparsecode_conv_admm.py:92]: `L1MinusL2Norm`? `L1NormMinusL2Norm`?

[the remark this answers is not in the document]
Comment (bwohlberg) [reply examples/scripts/sparsecode_conv_admm.py:92]: Hmm. I considered those before, but `L1MinusL2Norm` actually does seem preferable to `DiffL1L2Norm`.

**Pull request merged**